shared-bindings/analogio/AnalogFastIn.c

```diff
@@ -0,0 +1,197 @@
+ *
+ * The MIT License (MIT)
+ *
+ * SPDX-FileCopyrightText: Copyright (c) 2022 Lee Atkinson, MeanStride Technology, Inc.
+ * Taken from AnalogIn by Damien P. George
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <string.h>
+
+#include "shared/runtime/context_manager_helpers.h"
+#include "py/binary.h"
+#include "py/mphal.h"
+#include "py/nlr.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "shared-bindings/analogio/AnalogFastIn.h"
+#include "shared-bindings/util.h"
+
+
+
+//| class AnalogFastIn:
+//|     """Read analog voltage levels quickly using DMA Capture"""
+//|
+//|     def __init__(self, pin: microcontroller.Pin, buffer: ReadableBuffer, *, sample_rate: int = 500000) -> None:
+//|         """Use the AnalogFastIn on the given pin. Fill the given buffer from ADC read values at the supplied
+///         sample_rate.
+///
+//|         :param ~microcontroller.Pin pin: the pin to read from"""
+//|         :param ~circuitpython_typing.WriteableBuffer buffer: A buffer for samples
+//|         :param int sample_rate: The desired playback sample rate
+///
+//|     Usage::
+///
+//|       import board
+//|       import analogio
+//|       import array
+///
+//|       length = 1000
+//|       mybuffer = array.array("H", [0] * length)
+//|       fadc = analogio.AnalogFastIn(board.GP26, mybuffer)
+//|       fadc.capture()
+//|       fadc.deinit()
+//|       for i in range(length):
+//|           print(i, mybuffer[i])
+///
+///       (Future) The reference voltage varies by platform so use ``reference_voltage`` to read the configured setting.
+///       """
+//|         ...
+///
+STATIC void validate_rate(mp_float_t rate) {
+    if (rate < (mp_float_t)1.0f || rate > (mp_float_t)500000.0f) {
+        mp_raise_ValueError(translate("sample rate must be 1.0-500000.0 per second"));
+    }
+}
+
```

Comment:
Reuse an existing error message string here to save space. See `py/argcheck.c`. We might consider adding a new float range `mp_arg_validate...` routine there.